pkg/phlaredb/head.go

```diff
@@ -297,21 +305,61 @@ func (h *Head) convertSamples(ctx context.Context, r *rewriter, in []*profilev1.
 			// will copy it, if it is required to be retained.
 			LocationIDs: in[idxSample].LocationId,
 		}
+		for i := range stacktraces[idxSample].LocationIDs {
+			r.locations.rewriteUint64(&stacktraces[idxSample].LocationIDs[i])
+		}
 	}
+	appender := h.symbolDB.MappingWriter(stacktracePartition).StacktraceAppender()
+	defer appender.Release()
 
-	// ingest stacktraces
-	if err := h.stacktraces.ingest(ctx, stacktraces, r); err != nil {
-		return nil, err
+	if cap(stacktracesIds) < len(stacktraces) {
+		stacktracesIds = make([]uint32, len(stacktraces))
 	}
+	stacktracesIds = stacktracesIds[:len(stacktraces)]
+	defer uint32SlicePool.Put(stacktracesIds)
+
+	appender.AppendStacktrace(stacktracesIds, stacktraces)
+
+	h.metrics.sizeBytes.WithLabelValues("stacktraces").Set(float64(h.symbolDB.MemorySize()))
 
 	// reference stacktraces
 	for idxType := range out {
 		for idxSample := range out[idxType] {
-			out[idxType][idxSample].StacktraceID = uint64(r.stacktraces[int64(idxSample)])
+			out[idxType][idxSample].StacktraceID = uint64(stacktracesIds[int64(idxSample)])
+		}
+	}
+
+	return out
+}
+
+func StacktracePartitionFromProfile(lbls []phlaremodel.Labels, p *profilev1.Profile) uint64 {
+	return xxhash.Sum64String(stacktracePartitionKeyFromProfile(lbls, p))
+}
+
+func stacktracePartitionKeyFromProfile(lbls []phlaremodel.Labels, p *profilev1.Profile) string {
+	// take the first mapping (which is the main binary's file basename)
+	if len(p.Mapping) > 0 {
+		if filenameID := p.Mapping[0].Filename; filenameID > 0 {
+			filename := p.StringTable[filenameID]
+
+			// filter out invalid filenames like `[vsdo]` and `[vsyscall]`
+			// TODO: Ensure we handle windows paths correctly
+			if filepath.IsAbs(filename) {
+				return filepath.Base(filename)
+			}
+		}
+	}
+
+	// failing that look through the labels for the ServiceName
+	if len(lbls) > 0 {
+		for _, lbl := range lbls[0] {
+			if lbl.Name == phlaremodel.LabelNameServiceName {
+				return lbl.Value
+			}
 		}
 	}
```

Comment:
I'm not too sure but shouldn't we also accept file names that are not an absolute path?

Reply:
Copied some checks from google pprof package + added very basic sanitiser

I think we need a whole set of new limits. For example, number of mappings (main mappings), locations, unique stack traces per block, etc. plus size-based limitations, i.e: stack depth, function name max length, path map length and so on

Reply:
@korniltsev could you take a look please? Any ideas on what we should expect here are appreciated. At a minimum, it should be aligned with our own SDKs and clients

For context: the idea is to use this filename attribute as partitioning key of the stacktraces data set, ideally, the key should match the profiled program (its main binary), including all its versions. If the name is invalid or not present, we fallback to service_name attribute which may cause quite a significant duplication (if the same app has different values of the label)

Reply:
For our sdks it should be mostly ok because it is only golang we care about8ik,

node-js has no mappings 
pyroscope-rs ( ruby, python)
dotnet have no mappings 

go pprof only contains executable mappings and usually the first entry is the binary, however there are obviously ways to mess with it

-  -buildmode=pie

```
c000000000-c000400000 rw-p 00000000 00:00 0 
c000400000-c004000000 ---p 00000000 00:00 0 
55db19d0d000-55db19d65000 r-xp 00000000 103:03 6821170                   /home/korniltsev/trash/qwe
55db19d65000-55db19d8c000 r--p 00058000 103:03 6821170                   /home/korniltsev/trash/qwe
```
-   specifying arbitrary large .text address
```
go build -ldflags="-extldflags=-static -T 1099511623680" -o hello-static hello.go
```

Another thing that comes to mind is that generic name, may be used in different service 


/bin/grafana-agent or /bin/phlare running different components/targets may have different stacktraces and we may or may not keep partition key the same for them

Maybe we proactively blacklist something very common, like /bin/bash or /bin/python

If we ever add mappigns to ebpf, it may contain multiple binaries for multiple processes

Found these interesting mappings on my machine, not golang, so maybe just ignore them

```
4145b000-4145d000 r-xs 00000000 00:01 9225                               /memfd:/.glXXXXXX (deleted)
```


```
560b82d40000-560b82d43000 rw-p 00000000 00:00 0 
560b82d43000-560b82d44000 ---p 00000000 00:00 0 
560b82d44000-560b82d7f000 rwxp 00000000 00:00 0 << probably wasm or jit compiled js 
560b82d7f000-560b9fe80000 ---p 00000000 00:00 0 
560b9fe80000-560b9ffec000 r-xp 00000000 00:00 0 
560b9ffec000-560ba0000000 ---p 00000000 00:00 0 
560c06af4000-560c08902000 r--p 00000000 103:02 1993406                   /usr/lib/slack/slack
560c08903000-560c0f78e000 r-xp 01e0e000 103:02 1993406                   /usr/lib/slack/slack
560c0f78e000-560c0fced000 r--p 08c98000 103:02 1993406                   /usr/lib/slack/slack
560c0fced000-560c0fd6a000 rw-p 091f6000 103:02 1993406                   /usr/lib/slack/slack
```

If I run elf with linker manuallu, virtual memory layout changes

```
/lib64/ld-linux-x86-64.so.2 /bin/cat
```

```
cat /proc/466492/maps
55555604e000-55555606f000 rw-p 00000000 00:00 0                          [heap]
7f36cf858000-7f36d0100000 r--p 00000000 103:02 1968412                   /usr/lib/locale/locale-archive
7f36d0100000-7f36d0103000 rw-p 00000000 00:00 0 
7f36d0103000-7f36d012b000 r--p 00000000 103:02 1968212                   /usr/lib/x86_64-linux-gnu/libc.so.6
7f36d012b000-7f36d02c0000 r-xp 00028000 103:02 1968212                   /usr/lib/x86_64-linux-gnu/libc.so.6
7f36d02c0000-7f36d0318000 r--p 001bd000 103:02 1968212                   /usr/lib/x86_64-linux-gnu/libc.so.6
7f36d0318000-7f36d031c000 r--p 00214000 103:02 1968212                   /usr/lib/x86_64-linux-gnu/libc.so.6
7f36d031c000-7f36d031e000 rw-p 00218000 103:02 1968212                   /usr/lib/x86_64-linux-gnu/libc.so.6
7f36d031e000-7f36d034d000 rw-p 00000000 00:00 0 
7f36d034d000-7f36d034f000 rw-p 00000000 00:00 0 
7f36d034f000-7f36d0351000 r--p 00000000 103:02 1966228                   /usr/bin/cat
7f36d0351000-7f36d0355000 r-xp 00002000 103:02 1966228                   /usr/bin/cat
7f36d0355000-7f36d0357000 r--p 00006000 103:02 1966228                   /usr/bin/cat
7f36d0357000-7f36d0358000 r--p 00007000 103:02 1966228                   /usr/bin/cat
7f36d0358000-7f36d0359000 rw-p 00008000 103:02 1966228                   /usr/bin/cat
7f36d0359000-7f36d035d000 r--p 00000000 00:00 0                          [vvar]
7f36d035d000-7f36d035f000 r-xp 00000000 00:00 0                          [vdso]
7f36d035f000-7f36d0361000 r--p 00000000 103:02 1968200                   /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
```

All this virtual memory layouts may vary depending on architecture, I cant speak for arm.


Wrote a quick script to dump first executable entry for all processess
```
import os


def parse_one(pid):
    maps = open("/proc/%d/maps" % pid).read()
    for map in maps.splitlines():
        fields = map.split()
        # print(fields)
        if len(fields) < 6:
            continue
        if 'x' not in fields[1]:
            continue
        # print(map)
        return (map.strip(), fields[5], maps)
    return (None, None, maps)



for f in os.listdir("/proc"):
    try:
        pid = int(f)
    except:
        continue

    try:
        found, file, all = parse_one(pid)
        if found is not None:
            print(file)
    except:
        continue
```

a single node in dev cluster
<details>
/app/conntrack-exporter
/bin/agent
/bin/bash
/bin/m3dbnode
/bin/memcached_exporter
/bin/node_exporter
/bin/vault
/cloud_sql_proxy
/csi-node-driver-registrar
/dnsmasq-nanny
/gce-pd-csi-driver
/home/kubernetes/bin/kubelet
/home/kubernetes/bin/node-problem-detector
/ip-masq-agent
/keda
/kube-dns
/node-cache
/pause
/redis_exporter
/sbin/agetty
/sidecar
/usr/bin/bash
/usr/bin/calico-node
/usr/bin/cloud-backend-gateway
/usr/bin/containerd
/usr/bin/containerd-shim-runc-v2
/usr/bin/coreutils
/usr/bin/dbus-daemon
/usr/bin/dockerd
/usr/bin/dumb-init
/usr/bin/google_guest_agent
/usr/bin/google_osconfig_agent
/usr/bin/grafana-agent
/usr/bin/loki-canary
/usr/bin/metrics-instance-mapper
/usr/bin/nsenter
/usr/bin/promtail
/usr/bin/python3.11
/usr/bin/udevadm
/usr/lib/chromium/chrome
/usr/lib/systemd/systemd
/usr/lib/systemd/systemd-journald
/usr/lib/systemd/systemd-logind
/usr/lib/systemd/systemd-networkd
/usr/lib/systemd/systemd-resolved
/usr/local/bin/kube-proxy
/usr/local/bin/memcached
/usr/local/bin/node
/usr/local/bin/redis-server
/usr/local/bin/runsv
/usr/local/bin/runsvdir
/usr/sbin/chronyd
/usr/sbin/device_policy_manager
/usr/sbin/dnsmasq
/usr/sbin/sshd
/usr/share/grafana/bin/grafana-server
</details>

my machine

<details>
/home/korniltsev/Applications/Telegram/Telegram
/home/korniltsev/.asdf/installs/golang/1.19/packages/bin/gopls
/home/korniltsev/.asdf/installs/k9s/0.26.3/bin/k9s
/home/korniltsev/.asdf/installs/kubectl/1.24.2/bin/kubectl
/home/korniltsev/.asdf/installs/nodejs/16.18.0/bin/node
/home/korniltsev/.asdf/installs/python/3.11.0/bin/python3.11
/home/korniltsev/.asdf/installs/python/3.11.0/bin/python /home/korniltsev/PycharmProjects/pythonProject4/main.py 
/home/korniltsev/.cache/JetBrains/GoLand2023.1/tmp/GoLand/___1go_build_github_com_pyroscope_io_pyroscope_pkg_convert_pprof_qwe
/home/korniltsev/.cache/JetBrains/GoLand2023.1/tmp/GoLand/___7pyroscope_server_4040
/home/korniltsev/.config/Code/User/globalStorage/grafana.vscode-jsonnet/bin/jsonnet-language-server
/home/korniltsev/github/go-linux-amd64-bootstrap/bin/go
/home/korniltsev/github/go-linux-amd64-bootstrap/pkg/tool/linux_amd64/pprof
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/CLion/ch-1/221.5921.27/bin/clang/linux/clangd
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/CLion/ch-1/221.5921.27/bin/fsnotifier
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/CLion/ch-1/221.5921.27/jbr/bin/java
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/CLion/ch-1/221.5921.27.plugins/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/231.8770.71/bin/fsnotifier
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/231.8770.71/jbr/bin/java
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/231.8770.71/jbr/lib/jcef_helper
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/231.8770.71.plugins/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/231.8770.71/plugins/go-plugin/lib/dlv/linux/dlv
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/222.4459.20/bin/fsnotifier
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/222.4459.20/jbr/bin/java
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/222.4459.20.plugins/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/WebStorm/ch-0/222.4345.14/bin/fsnotifier
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/WebStorm/ch-0/222.4345.14/jbr/bin/java
/home/korniltsev/.local/share/JetBrains/Toolbox/apps/WebStorm/ch-0/222.4345.14/jbr/lib/jcef_helper
/home/korniltsev/trash/qwe
/memfd:/.glXXXXXX
/opt/google/chrome/chrome
/opt/google/chrome/chrome_crashpad_handler
/opt/google/chrome/nacl_helper
/opt/Surfshark/chrome_crashpad_handler
/opt/Surfshark/surfshark
Process finished with exit code 0
/usr/bin/akonadi_akonotes_resource
/usr/bin/akonadi_birthdays_resource
/usr/bin/akonadi_contacts_resource
/usr/bin/akonadi_control
/usr/bin/akonadi_followupreminder_agent
/usr/bin/akonadi_ical_resource
/usr/bin/akonadi_indexing_agent
/usr/bin/akonadi_maildir_resource
/usr/bin/akonadi_maildispatcher_agent
/usr/bin/akonadi_mailmerge_agent
/usr/bin/akonadi_migration_agent
/usr/bin/akonadi_newmailnotifier_agent
/usr/bin/akonadi_notes_agent
/usr/bin/akonadiserver
/usr/bin/cat
/usr/bin/dash
/usr/bin/dbus-daemon
/usr/bin/gjs-console
/usr/bin/gmenudbusmenuproxy
/usr/bin/gnome-keyring-daemon
/usr/bin/ibus-daemon
/usr/bin/ip
/usr/bin/kaccess
/usr/bin/kded5
/usr/bin/kdeinit5
/usr/bin/kglobalaccel5
/usr/bin/kgpg
/usr/bin/konsole
/usr/bin/korgac
/usr/bin/ksmserver
/usr/bin/kwalletd5
/usr/bin/less
/usr/bin/pipewire
/usr/bin/pipewire-media-session
/usr/bin/plasma_session
/usr/bin/pulseaudio
/usr/bin/python3.10
/usr/bin/startplasma-x11
/usr/bin/xembedsniproxy
/usr/bin/xsettingsd
/usr/bin/zsh
/usr/lib/bluetooth/obexd
/usr/libexec/at-spi2-registryd
/usr/libexec/at-spi-bus-launcher
/usr/libexec/dconf-service
/usr/libexec/evolution-addressbook-factory
/usr/libexec/evolution-calendar-factory
/usr/libexec/evolution-data-server/evolution-alarm-notify
/usr/libexec/evolution-source-registry
/usr/libexec/geoclue-2.0/demos/agent
/usr/libexec/goa-daemon
/usr/libexec/goa-identity-service
/usr/libexec/gvfs-afc-volume-monitor
/usr/libexec/gvfsd
/usr/libexec/gvfsd-fuse
/usr/libexec/gvfsd-http
/usr/libexec/gvfsd-metadata
/usr/libexec/gvfs-goa-volume-monitor
/usr/libexec/gvfs-gphoto2-volume-monitor
/usr/libexec/gvfs-mtp-volume-monitor
/usr/libexec/gvfs-udisks2-volume-monitor
/usr/libexec/ibus-engine-simple
/usr/libexec/ibus-extension-gtk3
/usr/libexec/ibus-memconf
/usr/libexec/ibus-portal
/usr/libexec/ibus-ui-gtk3
/usr/libexec/ibus-x11
/usr/libexec/pulse/gsettings-helper
/usr/libexec/tracker-miner-fs-3
/usr/libexec/xdg-desktop-portal
/usr/libexec/xdg-desktop-portal-gnome
/usr/libexec/xdg-desktop-portal-gtk
/usr/libexec/xdg-document-portal
/usr/libexec/xdg-permission-store
/usr/lib/slack/chrome_crashpad_handler
/usr/lib/slack/slack
/usr/lib/systemd/systemd
/usr/lib/x86_64-linux-gnu/libc.so.6
/usr/lib/x86_64-linux-gnu/libexec/baloorunner
/usr/lib/x86_64-linux-gnu/libexec/DiscoverNotifier
/usr/lib/x86_64-linux-gnu/libexec/kactivitymanagerd
/usr/lib/x86_64-linux-gnu/libexec/kdeconnectd
/usr/lib/x86_64-linux-gnu/libexec/kf5/kioslave5
/usr/lib/x86_64-linux-gnu/libexec/kf5/klauncher
/usr/lib/x86_64-linux-gnu/libexec/kf5/kscreen_backend_launcher
/usr/lib/x86_64-linux-gnu/libexec/xdg-desktop-portal-kde
/usr/sbin/mariadbd
/usr/share/code/chrome_crashpad_handler
/usr/share/code/code
</details>

Reply:
although it looks ok, I can't prove or promise it is actually ok, maybe I just don't know something

Reply:
Thank you for the very valuable input

> go pprof only contains executable mappings and usually the first entry is the binary, however there are obviously ways to mess with it

Basically, it means that even Go profiler may produce profiles without the main binary, correct? My understanding was that the tricks you listed only affect addresses but the main binary mapping is still present and its attributes like filename and build_id are still valid. I wonder how things are with PIC/PIE/ASLR with regard to profiling in general – I need to learn more and experiment with it.

When it comes to abuse, we should protect ourselves from (un-)intendedly malformed profiles, I believe

> Another thing that comes to mind is that generic name, may be used in different service
> /bin/grafana-agent or /bin/phlare running different components/targets may have different stacktraces and we may or may not keep partition key the same for them

Correct. I'm thinking about implementing a simple deduplication: at ingestion, override the partition key if most of the profile stack traces have already been written to a different partition. Imperfect but should work out

> Maybe we proactively blacklist something very common, like /bin/bash or /bin/python

That's a good idea, however might be challenging to maintain the list. It would be great if we could avoid this

> If we ever add mappigns to ebpf, it may contain multiple binaries for multiple processes

I guess we can create a stub mapping, given that we know its PID. Taking into account that we send symbolised profiles, addresses don't mean much for us as for now

Reply:
> node-js has no mappings
pyroscope-rs ( ruby, python)
dotnet have no mappings

No mapping at all, even a dummy one?

Reply:
> Basically, it means that even Go profiler may produce profiles without the main binary, correct?
no, I meant mappings with executable permissions https://github.com/golang/go/blob/master/src/runtime/pprof/proto.go#L705

> No mapping at all, even a dummy one?
No mapping at all